contracts/oracle/PCVOracle.sol

```diff
@@ -0,0 +1,396 @@
+    /// only callable by an illiquid pcv deposit that has previously been listed
+    /// in the PCV Oracle, because an oracle has to be set for the msg.sender.
+    /// this allows for lazy evaluation of the TWAPCV
+    /// @param pcvDelta the amount of PCV change in the venue
+    function updateIlliquidBalance(
+        int256 pcvDelta
+    )
+        public
+        onlyVoltRole(VoltRoles.ILLIQUID_PCV_DEPOSIT_ROLE)
+        isGlobalReentrancyLocked
+    {
+        _updateBalance(_getUsdPcvDelta(msg.sender, pcvDelta), false);
+        _afterActionHook();
+    }
+
+    /// ------------- Governor Only API -------------
+
+    /// @notice set the oracle for a given venue, used to normalize
+    /// balances into USD values, and correct for exceptional gains
+    /// and losses that are not properly reported by the PCVDeposit
+    function setVenueOracle(
+        address venue,
+        address newOracle
+    ) external onlyGovernor {
+        _setVenueOracle(venue, newOracle);
+    }
+
+    /// @notice add venues to the oracle
+    /// only callable by the governor
+    /// This locks system at level 1, because it needs to accrue
+    /// on the added PCV Deposits (that locks at level 2).
+    function addVenues(
+        address[] calldata venues,
+        address[] calldata oracles,
+        bool[] calldata isLiquid
+    ) external onlyGovernor globalLock(1) {
+        uint256 length = venues.length;
+        require(oracles.length == length, "PCVO: invalid oracles length");
+        require(isLiquid.length == length, "PCVO: invalid isLiquid length");
+        bool nonZeroBalances = false;
+        for (uint256 i = 0; i < length; ) {
+            require(venues[i] != address(0), "PCVO: invalid venue");
+            require(oracles[i] != address(0), "PCVO: invalid oracle");
+
+            // add venue in state
+            _setVenueOracle(venues[i], oracles[i]);
+            _addVenue(venues[i], isLiquid[i]);
+
+            uint256 balance = IPCVDepositV2(venues[i]).accrue();
+            if (balance != 0) {
+                nonZeroBalances = true;
+                // no need for safe cast here because balance is always > 0
+                _updateBalance(
+                    _getUsdPcvDelta(venues[i], int256(balance)),
+                    isLiquid[i]
+                );
+            }
+
+            unchecked {
+                ++i;
+            }
+        }
+        if (nonZeroBalances) _afterActionHook();
+    }
+
+    /// @notice remove venues from the oracle
+    /// only callable by the governor
+    /// This locks system at level 1, because it needs to accrue
+    /// on the added PCV Deposits (that locks at level 2).
+    function removeVenues(
+        address[] calldata venues,
+        bool[] calldata isLiquid
+    ) external onlyGovernor globalLock(1) {
+        uint256 length = venues.length;
+        require(isLiquid.length == length, "PCVO: invalid isLiquid length");
```

Comment:
convention we use throughout the codebase for revert messages is "FullContractName: reason"